src/bufr/BufrParser/Query/QueryRunner.cpp

```diff
@@ -382,7 +382,7 @@ namespace bufr {
                             for (int pathIdx = currentPath.size() - 1; pathIdx >= 0; --pathIdx) {
                                 returnNodeIdx = dataProvider_->getLink(
                                         dataProvider_->getJmpb(currentPath[pathIdx]));
-                                lastNonZeroReturnIdx = currentPathReturns.size() - pathIdx;
+                                lastNonZeroReturnIdx = currentPathReturns.size() - 1 - pathIdx;
```

Comment:
So this code should fortunately disappear when I do the Refactoring Part 2 to avoid all this kind of logic (data will be modeled as a node tree). Refactoring Part 1 does this with the Meta data and it helps clean things up a lot... I don't remember how this particular change came about, however. You have to be very careful here as this code is extremely finicky.

Reply:
@srherbener shouldn't the last equation be (3-(1+2)) = 0 ?  If you reduce the expression to one without paranthesis, what you have would be 3 - 1 + 2.

Reply:
I think you might be missing a sign change: 3 - (1 - 2) = 3 + (-1 + 2) = 4

Reply:
I'm not really sure this is an issue as the associativity is part of the definition of the operator itself (compilers should all do the exact same thing as far as I understand). Have you run into an actual situation where a compiler does it the wrong way? Please take a look at https://en.cppreference.com/w/cpp/language/operator_precedence

Reply:
@rmclaren thanks for the information. I wasn't aware of the spec that says the compiler must execute left-to-right order for subtraction. I tend to go for "defensive" programming especially when it's simple and quick to do so. To answer your question, I haven't seen a case where the compiler went the "wrong" way, but I tend to put in the parentheses to explicitly specify the order.